### PR TITLE
gb-studio: init at 3.1.0

### LINKED ISSUE #185564: GB Studio
**Project description**
a quick and easy to use retro adventure game creator for Game Boy

**Metadata**

* homepage URL: https://www.gbstudio.dev/
* source URL: https://github.com/chrismaltby/gb-studio
* license: mit https://github.com/chrismaltby/gb-studio/blob/develop/LICENSE
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by musjj**
- maintainers: add musjj
- gb-studio: init at 3.1.0

## Description of changes

GB Studio: A quick and easy to use drag and drop retro game creator for your favourite handheld video game system.

Just a simple wrapper around the pre-built AppImage executable for now. I really wanted to build the package from source, but I gave up after spending two whole days fighting with Yarn, Electron and RPATH.

closes: #185564

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pbsds** [comment]
```
gb-studio> building '/nix/store/kll1z65rxgr5pvrr5qr85krxd3hcygph-gb-studio-3.1.0.drv'
gb-studio> mv: cannot stat '/nix/store/ssbxl2490km2plqbdr5wrvlnzzw8sxjx-gb-studio-3.1.0/bin/gb-studio-3.1.0': No such file or directory
```

**Comment by ASHGOLDOFFICIAL** [comment]
I don't know how to install desktop file and icons. To do so, I need to override some phase but every phase is needed and is hard to recreate. Anyway, it builds (sometimes). But I don't know how to `npm start` after that. So I'll leave my attempt for someone else.

My progress:
```nix
{ lib
, mkYarnPackage
, fetchFromGitHub
, makeDesktopItem
, imagemagick
}:

mkYarnPackage rec {
  pname = "gb-studio";
  version = "3.2.1";

  src = fetchFromGitHub {
    owner = "chrismaltby";
    repo = "gb-studio";
    rev = "refs/tags/v${version}";
    hash = "sha256-hF8aDTg7v9p0d9AUE2NNhQACgggMEbi5afzK1GqUHx0=";
  };

  # doDist = false;

  nativeBuildInputs = [
    imagemagick
  ];

  # installPhase = let
  #   desktopItem = makeDesktopItem {
  #     name = pname;
  #     desktopName = "GB Studio";
  #     comment = "Visual retro game maker";
  #     genericName = "GB Studio";
  #     exec = "${meta.mainProgram} %U";
  #     icon = pname;
  #     type = "Application";
  #     startupNotify = true;
  #     categories = [ "Development" "Building" "IDE" ];
  #   };
  # in ''
  #   runHook preInstall
  #
  #   pushd ./deps/gb-studio
  #   for size in 64 128 256 512 1024; do
  #     mkdir -p $out/share/icons/hicolor/"$size"x"$size"/
  #     convert -resize "$size"x"$size" ./src/assets/app/icon/app_icon.png \
  #        $out/share/icons/hicolor/"$size"x"$size"/${pname}.png
  #   done
  #   popd
  #
  #   install -Dm 644 ${desktopItem}/share/applications/*.desktop \
  #     -t $out/share/applications
  #
  #   runHook postInstall
  # '';

  meta = {
    description = "A quick and easy to use drag and drop retro game creator for your favourite handheld video game system";
    homepage = "https://www.gbstudio.dev/";
    downloadPage = "https://chrismaltby.itch.io/gb-studio";
    changelog = "https://github.com/chrismaltby/gb-studio/releases/tag/v${version}";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ musjj ];
    mainProgram = pname;
    platforms = lib.platforms.linux;
  };
}
```

**Comment by pbsds** [comment]
Try the copyDesktopItems hook

File: pkgs/development/tools/gb-studio/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      genericName = "GB Studio";
+      exec = "${meta.mainProgram} %U";
+      icon = pname;
+      type = "Application";
+      startupNotify = true;
+      categories = [ "GNOME" "GTK" "Utility" ];
+    };
+    # bash
+  in ''
+    mv "$out/bin/${pname}-${version}" "$out/bin/${pname}"
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:31]:
```suggestion
```

File: pkgs/development/tools/gb-studio/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
+{ lib, appimageTools, fetchurl, makeDesktopItem }:
+
+appimageTools.wrapType2 rec {
+  pname = "gb-studio";
+  version = "3.1.0";
+
+  src = fetchurl {
+    url =
+      "https://github.com/chrismaltby/${pname}/releases/download/v${version}/${pname}-linux.AppImage";
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:9]:
```suggestion
      "https://github.com/chrismaltby/gb-studio/releases/download/v${version}/gb-studio-linux.AppImage";
```

File: pkgs/development/tools/gb-studio/default.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    mkdir -p $out/share/applications
+    cp ${desktopItem}/share/applications/*.desktop $out/share/applications
+  '';
+
+  meta = with lib; {
+    description =
+      "A quick and easy to use drag and drop retro game creator for your favourite handheld video game system";
+    homepage = "https://www.gbstudio.dev/";
+    changelog =
+      "https://github.com/chrismaltby/${pname}/releases/tag/v${version}";
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:41]:
```suggestion
      "https://github.com/chrismaltby/gb-studio/releases/tag/v${version}";
```

File: pkgs/development/tools/gb-studio/default.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    # bash
+  in ''
+    mv "$out/bin/${pname}-${version}" "$out/bin/${pname}"
+    mkdir -p $out/share/applications
+    cp ${desktopItem}/share/applications/*.desktop $out/share/applications
+  '';
+
+  meta = with lib; {
+    description =
+      "A quick and easy to use drag and drop retro game creator for your favourite handheld video game system";
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:38]:
```suggestion
      "Quick and easy to use drag and drop retro game creator for your favourite handheld video game system";
```

File: pkgs/development/tools/gb-studio/default.nix [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      categories = [ "GNOME" "GTK" "Utility" ];
+    };
+    # bash
+  in ''
+    mv "$out/bin/${pname}-${version}" "$out/bin/${pname}"
+    mkdir -p $out/share/applications
+    cp ${desktopItem}/share/applications/*.desktop $out/share/applications
+  '';
+
+  meta = with lib; {
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:36]:
```suggestion
  meta = {
```

tracking issue: #292468

File: pkgs/development/tools/gb-studio/default.nix [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    description =
+      "A quick and easy to use drag and drop retro game creator for your favourite handheld video game system";
+    homepage = "https://www.gbstudio.dev/";
+    changelog =
+      "https://github.com/chrismaltby/${pname}/releases/tag/v${version}";
+    mainProgram = pname;
+    license = licenses.mit;
+    platforms = [ "x86_64-linux" ];
+    sourceProvenance = with sourceTypes; [ binaryBytecode ];
+    maintainers = with maintainers; [ musjj ];
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:46]:
```suggestion
    license = lib.licenses.mit;
    platforms = [ "x86_64-linux" ];
    sourceProvenance = with lib.sourceTypes; [ binaryBytecode ];
    maintainers = with lib.maintainers; [ musjj ];
```

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -3846,6 +3846,8 @@ with pkgs;
 
   gams = callPackage ../tools/misc/gams (config.gams or {});
 
+  gb-studio = callPackage ../development/tools/gb-studio { };
```
Comment (eclairevoyant) [inline pkgs/top-level/all-packages.nix:3849]: please move this to `pkgs/by-name` and remove this line

File: pkgs/development/tools/gb-studio/default.nix [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      icon = pname;
+      type = "Application";
+      startupNotify = true;
+      categories = [ "GNOME" "GTK" "Utility" ];
+    };
+    # bash
+  in ''
+    mv "$out/bin/${pname}-${version}" "$out/bin/${pname}"
+    mkdir -p $out/share/applications
+    cp ${desktopItem}/share/applications/*.desktop $out/share/applications
```
Comment (eclairevoyant) [inline pkgs/development/tools/gb-studio/default.nix:33]:
```suggestion
    install -Dm444 ${desktopItem}/share/applications/*.desktop -t $out/share/applications
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Closing because of unresponsive author

**Comment by pbsds** [comment]
since when did we start closing stale PRs?

File: pkgs/development/tools/gb-studio/default.nix (hunk 8 above)
Comment (eclairevoyant) [reply pkgs/development/tools/gb-studio/default.nix:33]: missed the earlier comment, oops

**Pull request closed without merging**